### DISCUSSION TITLE
list object v2 - MaxKey, KeyCount , EncodingType

### DISCUSSION THREAD

**Discussion by shirady** [aws/aws-cli#7487] [Q&A]
Hi,
A couple of questions about object list V2:
1. Should `MaxKey` be equal to the integer that was sent with option `--max-items `?
2. Should `KeyCount` be equal to the number of keys in the reply inside content?
3. Should `EncodingType` be in the response (although it was not specified in the command)?
4. How can we differentiate between a request from the client if he wants the response decoded?
As I found the query params look the same when we send the command with or without the flag `--encoding-type url`.

For example I run this command in AWS CLI:
`aws s3api list-objects-v2 --bucket shira-test-s3-3 --max-items 2 --debug`
The response was with 2 keys only, but the `MaxKey`is still 1,000, and the `KeyCount` is the number of the keys in the bucket (4), but not the number of the keys in the content.  `EncodingType` is url although we didn't specify it in the command.

```
<ListBucketResult
	xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
	<Name>shira-test-s3-3</Name>
	<Prefix></Prefix>
	<KeyCount>4</KeyCount>
	<MaxKeys>1000</MaxKeys>
	<EncodingType>url</EncodingType>
	<IsTruncated>false</IsTruncated>
	<Contents>
		<Key>asdf%2Bb</Key>
		<LastModified>2022-11-28T07:46:09.000Z</LastModified>
		<ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag>
		<Size>0</Size>
		<StorageClass>STANDARD</StorageClass>
	</Contents>
	<Contents>
		<Key>foo%2B1%3Abar</Key>
		<LastModified>2022-11-28T07:46:08.000Z</LastModified>
		<ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag>
		<Size>0</Size>
		<StorageClass>STANDARD</StorageClass>
	</Contents>
	<Contents>
		<Key>foo%3Abar%3Axyzzy</Key>
		<LastModified>2022-11-28T07:46:08.000Z</LastModified>
		<ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag>
		<Size>0</Size>
		<StorageClass>STANDARD</StorageClass>
	</Contents>
	<Contents>
		<Key>quux+ab%3Athud</Key>
		<LastModified>2022-11-28T07:46:07.000Z</LastModified>
		<ETag>&quot;d41d8cd98f00b204e9800998ecf8427e&quot;</ETag>
		<Size>0</Size>
		<StorageClass>STANDARD</StorageClass>
	</Contents>
</ListBucketResult>'
```

**Comment by tim-finnigan** [comment]
Hi @shirady thanks for reaching out. Here is the [AWS CLI documentation](https://docs.aws.amazon.com/cli/latest/reference/s3api/list-objects-v2.html) for the `list-objects-v2` command and [here is the   documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html) for the underlying S3 ListObjectsV2 API. For more reference on `MaxKeys`, `KeyCount`, and `EncodingType`:

<img width="713" alt="image" src="https://user-images.githubusercontent.com/87778557/205711863-d9ca1605-78fb-4abc-918f-f898056f574f.png">
<img width="664" alt="image" src="https://user-images.githubusercontent.com/87778557/205711922-8a792610-670b-433e-9405-c514a900ab1f.png">

The [--max-items parameter](https://docs.aws.amazon.com/cli/latest/userguide/cli-usage-pagination.html#cli-usage-pagination-maxitems) is used for client-side pagination, so it doesn't affect the underlying API request.
Please let us know if you had any follow up questions or you'd like more clarification on anything.

**Comment by shirady** [comment]
Hi @tim-finnigan,
Thank you for your response, I still have open questions:
1. How the client can limit the `MaxKey`? (as you answered the flag `--max-items` is not relevant). 
2. In a case where the response has common prefixes - it is not mentioned in the documentation - What should be the `keyCount`?
3. What is the standard that is used for encoding? (It comes from a certain RFC)
For example, I saw that the encoded response of space (SP) is '+' but it can be also '%20'.
